### PR TITLE
Add debug switch in config & extra specs

### PR DESCRIPTION
**Pull request by maigl**
This add a convenient 'debug-switch' which essentially adds a `set -x` in the cloud init script.

This switch can be toggled in the provider config, but probably the more interesting use-case is to toggle temporarily this via the `extra-specs`.

### remarks
It feels weird that the actual cloud-init script is still in a package in `garm` and has to be changed in the dependencies. Shouldn't this be code in the provider? Anyway, this PR probably would ne a corresponding change in garm itself.

What do you think?

<sub>Michael Kuhnt <michael.kuhnt@mercedes-benz.com> Mercedes-Benz Tech Innovation GmbH ([ProviderInformation](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md))
</sub>

### CODE REVIEW

**Comment by gabriel-samfira** [comment]
I think we can benefit from a separate package for the whole external provider framework. This should include the userdata along with a number of utility functions that gives us granular co trol over how we generate userdata.

I am currently OOF but will look into separating that stuff from garm itself. It doesn't really fit well there either. This package:

https://github.com/cloudbase/garm/tree/main/runner/providers/external/execution

Along with:

https://github.com/cloudbase/garm/tree/main/cloudconfig

Can be added to a new external package.


As to this PR, it looks good, but I think we should wait for the split, and just consume the new package. We also need to document that the debug option may leak secrets in the instance console/logs.

File: vendor/github.com/cloudbase/garm/params/params.go [hunk 1]
```diff
@@ -222,7 +222,9 @@ type BootstrapInstance struct {
 }
 
 type UserDataOptions struct {
-	DisableUpdatesOnBoot bool `json:"disable_updates_on_boot"`
+	DisableUpdatesOnBoot bool     `json:"disable_updates_on_boot"`
+	EnableBootDebug      bool     `json:"enable_boot_debug"`
```
Comment (gabriel-samfira) [inline vendor/github.com/cloudbase/garm/params/params.go:226]: This does not seem to be in any commit of garm. We normally don't manually change the contents of the `vendor` folder. It leads to inconsistencies.

But this should be addressed when we split the provider stuff from garm.

**Comment by gabriel-samfira** [comment]
In the process of creating that `garm-provider-common` repo with the packages we discussed. The first iteration will have the cloud config and external provider stuff. We can add stuff to it later. But it should make things easier to maintain.

**Comment by gabriel-samfira** [comment]
We've created: https://github.com/cloudbase/garm-provider-common

This package holds most of the functionality needed by external providers. I will update the OpenStack provider to use it, soon. You will most likely have to rebase your changes and fix any conflicts.

I took the liberty of adding the debug flag you proposed here, in the `garm-provider-common` repo.  Check out the [cloudconfig](https://github.com/cloudbase/garm-provider-common/blob/main/cloudconfig/util.go) package for details. It adds some nice new functionality, including the ability to override the default template used to install the runner.

Still needs testing, but wanted to give you a heads up.

**Comment by gabriel-samfira** [comment]
I've updated this provider to use [garm-provider-common](https://github.com/cloudbase/garm-provider-common). Could you rebase when you have the time? We can then review this PR.

**Comment by gabriel-samfira** [comment]
apropos, one potentially interesting feature added after this move, is the ability to [override the default installation script](https://github.com/cloudbase/garm-provider-common#userdata), supply additional template context and inject additional pre-install scripts (which will run in alphabetical order).

**Comment by maigl** [comment]
I rebased it ..

**Comment by gabriel-samfira** [comment]
Thanks! Will have a look later today!

**Comment by gabriel-samfira** [comment]
Actually, this is tiny. And it looks great. Merging. Thanks for the patience while I split the code from GARM :smile: .

**Pull request merged**